### PR TITLE
Make k-means init and EMA dead code expiry safe for distributed training

### PR DESCRIPTION
**Pull request by falkaer**
Hi,

I've been using this code and noticed that the k-means init and EMA dead code expiry options are not safe for use with DDP as they use vectors from each batch on separate replicas leading to each replica having different codebooks.

I solved this by making a distributed sampling function to randomly sample vectors uniformly from all batches in the DDP such that all replicas end up with the same sample consisting of items from all batches. I also introduced all_reduce calls in the k-means function such that samples from all replicas participate in the initialization.

I've verified that these changes lead to all codebooks remaining identical at all steps when using DDP. Let me know if there are any changes you need to accept.

Best regards,
Kenny

### CODE REVIEW

**Comment by lucidrains** [comment]
this looks great! thank you!

File: vector_quantize_pytorch/vector_quantize_pytorch.py [hunk 1]
```diff
@@ -130,19 +189,16 @@ def init_embed_(self, data):
...
         self.initted.data.copy_(torch.Tensor([True]))
 
     def replace(self, samples, mask):
-        modified_codebook = torch.where(
-            mask[..., None],
-            sample_vectors(samples, self.codebook_size),
-            self.embed
-        )
-        self.embed.data.copy_(modified_codebook)
+        self.embed.data[mask] = self.sample_fn(samples, mask.sum().item())
```
Comment (netw0rkf10w) [inline vector_quantize_pytorch/vector_quantize_pytorch.py:201]: Isn't `torch.count_nonzero(mask)` supposed to be faster than `mask.sum().item()` here? @lucidrains @falkaer
